truffle/src/main/java/org/jruby/truffle/core/exception/NameErrorLayout.java

```diff
@@ -0,0 +1,38 @@
+/*
+ * Copyright (c) 2016 Oracle and/or its affiliates. All rights reserved. This
+ * code is released under a tri EPL/GPL/LGPL license. You can use it,
+ * redistribute it and/or modify it under the terms of the:
+ *
+ * Eclipse Public License version 1.0
+ * GNU General Public License version 2
+ * GNU Lesser General Public License version 2.1
+ */
+package org.jruby.truffle.core.exception;
+
+import com.oracle.truffle.api.object.DynamicObject;
+import com.oracle.truffle.api.object.DynamicObjectFactory;
+import com.oracle.truffle.api.object.dsl.Layout;
+import com.oracle.truffle.api.object.dsl.Nullable;
+import org.jruby.truffle.language.backtrace.Backtrace;
+
+@Layout
+public interface NameErrorLayout extends ExceptionLayout {
+
+    DynamicObjectFactory createNameErrorShape(
+        DynamicObject logicalClass,
+        DynamicObject metaClass);
+
+    DynamicObject createNameError(
+        DynamicObjectFactory factory,
+        Object message,
+        @Nullable Backtrace backtrace,
+        Object name,
+        @Nullable Object receiver);
```

Comment:
Well, one trick is that if you subclass only adds fields you can pass all the fields to each layout and the super-layout will ignore those it doesn't use. Probably not a good idea to rely on this.